Add sea-surface height

einola opened the pull request:
# Add sea-surface height
Fixes #538 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [ ] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR adds the contribution of the sea-surface-height gradient to both the mEVP and BBM codes (through VPCGDynamicsKernel and BrittleDynamicsKernel). SSH is read in from IOceanBoundary and passed to the dynamical core. It can also be output through ConfigOutput as "ssh".

This PR is a repeat of PR #662 with some fixes to make sure the integration test runs.

---
# Test Description

No unit test was implemented for this feature, and no quantitative test exists. Qualitative testing shows that setting a fixed sea-surface-height gradient gives drift speed close to back-of-the-envelope estimates (13 cm/s modelled, 20 cm/s from back-of-the-envelope).

---
# Documentation Impact

None

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

winzerle committed:
Added the dG(0) variable SeasurfaceHeight.

This variable has to be set from outside.

CGDynamicsKernel::prepareIteration will compute the gradient of SeasurfaceHeight and stores it as CG-Vectors uGradSeasurfaceHeight and vGradSeasurfaceHeight ins CGDynamicsKernel. VPCGDynamicsKernel is using it in main update loop.

Before merging we should activate a ice-only testcase.

einola committed:
Plumbing to move ssh around

Adds ssh to IDynamics and IOceanBoundary classes and all the derived
classes. Also modifies TOPAZOcn_test.cpp and topaz_test_data.py to test
ssh, as well as changing era5_topaz4_maker.py to include and interpolate
 ssh.

einola committed:
Include SSH in BBM

einola committed:
Add a sensible name for SSH in output

And that is "ssh" (instead of "SSH").

einola committed:
Headers and formatting

... to appease clang-format and the PR template

einola committed:
Clang format fixes

... which I missed on the last commit.

einola committed:
More clang-format fixes

einola committed:
You guessed it!

einola committed:
Fix renaming/refactoring mistake

einola committed:
Fix a mistake where ssh was a VField not HField

As pointed out by Tim when reviewing the PR.

timspainNERSC committed:
Allow forcing fields to be absent from files.

timspainNERSC committed:
Handle absent sea surface height forcing data.

timspainNERSC committed:
Use gridnames in TOPAZOcean.

timspainNERSC committed:
formatting

einola committed:
Merge remote-tracking branch 'origin/develop' into issue538_seasurfaceheight_pure

I also had to address some minor conflicts. The root of those was that
we'd defined the gravity constant, g, in two places.

einola committed:
Minor clean up of degrees and radians conversion

Move deg2rad to constants.hpp and use this for the radians function.
Define also rad2deg and use in the degrees function ... but I still need
 to rewrite rad2deg as a proper hex float.

einola committed:
Bugfix for era5_topaz4_test_data.py

Strangely, time step two of the TOPAZ file was always NaNs. I don't
understand why the model still worked.

einola committed:
Add the ssh field to the integration test

Derived from the TOPAZ4 data, like the rest.

einola commented on core/src/include/constants.hpp:
```diff
@@ -31,6 +31,13 @@ const double Tt = 273.16;
+
+// radians to degrees
+// TODO: Rewrite as a hex float
+const double rad2deg = 1. / deg2rad;
```
```suggestion
const double rad2deg = 0x1.ca5dc1a63c1f8p+5;
```

I looked it up on the internet (https://observablehq.com/@jrus/hexfloat)

einola committed:
Update rad2deg in constants.hpp

I converted 180/pi to a hex float here: https://observablehq.com/@jrus/hexfloat

einola commented on core/src/include/constants.hpp:
```diff
@@ -31,6 +31,13 @@ const double Tt = 273.16;
+const double deg2rad = 0x1.1df46a2529d39p-6;
+
+// radians to degrees
+// TODO: Rewrite as a hex float
```
```suggestion
```

einola committed:
Remove a TODO in constants.hpp

This is a follow-up on the previous commit (should have used git commit-amend).

timspainNERSC commented on core/src/modules/include/ProtectedArrayNames.ipp:
```diff
@@ -40,3 +40,4 @@
     { "sss_slab", "SLAB_SSS" }, // Slab ocean surface salinity PSU
     { "qdw", "SLAB_QDW" }, // Slab ocean temperature nudging heat flux, W m⁻²
     { "fdw", "SLAB_FDW" }, // Slab ocean salinity nudging water flux, kg s⁻¹ m⁻²
+{ "ssh", "SSH" }, // Slab ocean salinity nudging water flux, kg s⁻¹ m⁻²
```
Align the indenting, even though its not processed by clang-format.

timspainNERSC replied:
Align the indenting, even though its not processed by clang-format.

timspainNERSC commented on dynamics/src/CGDynamicsKernel.cpp:
```diff
@@ -36,6 +36,9 @@ void CGDynamicsKernel<DGadvection>::initialise(
     cgH.resize_by_mesh(*smesh);
     cgA.resize_by_mesh(*smesh);
 
+    uGradSeasurfaceHeight.resize_by_mesh(*smesh);
+    vGradSeasurfaceHeight.resize_by_mesh(*smesh);
```
Perhaps x & y, rather than u & v?

timspainNERSC commented on dynamics/src/include/VPCGDynamicsKernel.hpp:
```diff
@@ -124,7 +127,9 @@ template <int DGadvection> class VPCGDynamicsKernel : public CGDynamicsKernel<DG
                         * (FAtm * absatm * uAtmos(i) + // atm forcing
                             FOcean * absocn * SC * uOcean(i)) // ocean forcing
-                    + params.rhoIce * cgH(i) * params.fc * vOcnRel // cor + surface
+                    - params.rhoIce * cgH(i) * params.fc * u(i) // Coriolis
```
Is `u(i)` correct here? I would expect the Coriolis term to take the other component of the 2D velocity.

einola replied:
Nice catch!

einola replied:
This is fixed now. Can you re-review @timspainNERSC ?

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/ConfiguredOcean.cpp:
```diff
@@ -96,6 +96,10 @@ void ConfiguredOcean::setData(const ModelState::DataMap& ms)
     tf = Module::getImplementation<IFreezingPoint>()(sssExt[0]);
     cpml = Water::rho * Water::cp * mld[0];
 
+    /* It's only the SSH gradient which has an effect, so being able to sett a constant SSH is
```
Det er bare 'set' på engelsk.

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/FluxConfiguredOcean.cpp:
```diff
@@ -78,6 +78,10 @@ void FluxConfiguredOcean::setData(const ModelState::DataMap& ms)
     tf = Module::getImplementation<IFreezingPoint>()(sss[0]);
     cpml = Water::rho * Water::cp * mld[0];
+
+    /* It's only the SSH gradient which has an effect, so being able to sett a constant SSH is
```
It's just 'set' in English.

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp:
```diff
@@ -57,14 +58,19 @@ void TOPAZOcean::configure()
 {
     // TODO: Get more authoritative names for the forcings
-    std::set<std::string> forcings = { "sst", "sss", "mld", "u", "v" };
+    std::set<std::string> forcings = { sstName, sssName, "mld", uName, vName, sshName };
```
Since `mldName` was added to `gridNames.hpp` it can be used here, too.

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/UniformOcean.cpp:
```diff
@@ -24,6 +24,10 @@ void UniformOcean::setData(const ModelState::DataMap& ms)
     cpml = Water::rho * Water::cp * mld[0];
     qio = qio0;
+
+    /* It's only the SSH gradient which has an effect, so being able to sett a constant SSH is
```
Another 'sett'.

timspainNERSC reviewed (changes requested):
Formatting, Icelandicisms and naming. Also, the correct component in the Coriolis term?

einola committed:
Minor changes after PR review

This commit addresses the minor changes Tim asked for in his review.
This includes re-naming variables and correcting spelling mistakes.

einola committed:
Fix sign error in mEVP solver

As per Tim's comment on PR #755. Also some mino cleaning of the mEVP
momentum solver code.

einola commented:
Icelandicisms is a new word for misspellings :D And it's difficult to spell, which is definitely a plus!

einola committed:
Merge remote-tracking branch 'origin/develop' into issue538_seasurfaceheight_pure

einola committed:
Merge remote-tracking branch 'origin/develop' into issue538_seasurfaceheight_pure

timspainNERSC reviewed (approved):
I think that all looks good, now
